Add CA and leaf cert validations

jondonas opened the pull request:
Adds validations specified here: https://github.com/notaryproject/notaryproject/blob/main/signature-specification.md#certificate-requirements

shizhMSFT reviewed (commented):
General golang styling comments:
- Document all exported methods
- Coding blocks like
  ```go
  if err := doSomeThing(); err != nil {
    return err
  }
  return nil
  ```
  can be simplified as
  ```go
  return doSomeThing()
  ```
- Acronyms should be all caps. e.g. `CA`, `EKU`, `OID` over `Ca`, `Eku`, `Oid`.
- Add comment and references if the code block is hard to be understood by others.

priteshbandi commented on x509/cert_validations.go:
```diff
@@ -0,0 +1,166 @@
+
+// VerifyCACertificate validates root/intermediate certificates according to this spec:
+// https://github.com/notaryproject/notaryproject/blob/main/signature-specification.md#root-and-intermediate-ca-certificates
+func VerifyCACertificate(cert *x509.Certificate, expectedPathLen int) error {
+	if err := verifyCABasicConstraints(cert, expectedPathLen); err != nil {
+		return err
+	}
+	return verifyCAKeyUsage(cert)
+}
+
+// VerifyCodeSigningCertificate validates code-signing certificates according to this spec:
+// https://github.com/notaryproject/notaryproject/blob/main/signature-specification.md#leaf-certificates
+func VerifyCodeSigningCertificate(cert *x509.Certificate) error {
+	return verifyCertificate(cert, x509.ExtKeyUsageCodeSigning)
+}
+
+// VerifyTimeStampingCertificate validates time-stamping certificates according to this spec:
+// https://github.com/notaryproject/notaryproject/blob/main/signature-specification.md#leaf-certificates
+func VerifyTimeStampingCertificate(cert *x509.Certificate) error {
+	return verifyCertificate(cert, x509.ExtKeyUsageTimeStamping)
+}
+
```
1. Can we please update [`ValidateCertChain`](https://github.com/notaryproject/notation-core-go/blob/main/x509/cert.go#L49) method with this validations?
2. If we do 1st then we dont need to make these methods public.
3. Should we move `ValidateCertChain` from `certs.go` to `cert_validations.go`

jondonas commented on x509/cert_validations.go:
```diff
@@ -0,0 +1,207 @@
+	return nil
+}
+
+func validateKeyUsagePresent(cert *x509.Certificate) error {
```
Just for presence yes, but that doesn't tell us KU is marked critical. We need to pull the specific Extension to check that

priteshbandi reviewed (approved):
LGTM

jondonas commented:
> General golang styling comments:
> 
>     * Document all exported methods
> 
>     * Coding blocks like
>       ```go
>       if err := doSomeThing(); err != nil {
>         return err
>       }
>       return nil
>       ```
>       
>       
>           
>             
>           
>       
>             
>           
>       
>           
>         
>       can be simplified as
>       ```go
>       return doSomeThing()
>       ```
> 
>     * Acronyms should be all caps. e.g. `CA`, `EKU`, `OID` over `Ca`, `Eku`, `Oid`.
> 
>     * Add comment and references if the code block is hard to be understood by others.



> General golang styling comments:
> 
>     * Document all exported methods
> 
>     * Coding blocks like
>       ```go
>       if err := doSomeThing(); err != nil {
>         return err
>       }
>       return nil
>       ```
>       
>       
>           
>             
>           
>       
>             
>           
>       
>           
>         
>       can be simplified as
>       ```go
>       return doSomeThing()
>       ```
> 
>     * Acronyms should be all caps. e.g. `CA`, `EKU`, `OID` over `Ca`, `Eku`, `Oid`.
> 
>     * Add comment and references if the code block is hard to be understood by others.

Addressed all these comments

priteshbandi reviewed (approved):
LGTM.
